### ISSUE TITLE
Autoselection of device

### ISSUE THREAD

**Issue by matsekberg** [arduino/Arduino#1481]
Enable the use of some meta-comment in the code so that the appropriate board is automatically selected for that code.

**Comment by ffissore** [comment]
Wontfix, but Arduino Create will provide a similar feature

**Comment by Avamander** [comment]
Um, what if we wish to use offline IDE and get this feature?

**Comment by Avamander** [comment]
What if we wish to use standalone Arduino IDE? We won't have this feature any more.

**Comment by ffissore** [comment]
I'm sorry but you'll have to code it on your own or gather traction by some other hacker so to raise the priority of this feature

**Comment by JamesNewton** [comment]
Expanded on this idea in https://github.com/arduino/Arduino/issues/5308
I think it has greater use with esp 8266

**Comment by bcraigie** [comment]
It is pretty much a no-brainer and a must-have that the board type is included in the INO file and I am adding my name to this to gain traction. :-)

**Comment by hachi** [comment]
@ffissore when is this feature coming? Been two years now.

**Comment by hachi** [comment]
@ffissore can we reopen this? it seems like arduino create is an entirely different product, and one shouldn't have to switch paradigms to get such a rudimentary feature. Two years later and people still want it in the IDE.

**Comment by ffissore** [comment]
Sorry you've missed the new @hachi but I no longer work for Arduino https://groups.google.com/a/arduino.cc/forum/#!msg/developers/YJLX6AZHem4/kpsLef7XBAAJ

**Comment by timkoers** [comment]
If you need any one to give it a try, count me in!
I'll see if I can create something.
If you can give a few examples about how you'd like the syntax to be and where ether you'd like a preferences.txt file or add it to a header in the main .ino file, might give it a try.

**Comment by JamesNewton** [comment]
@timkoers A few thoughts:

1. If there is a preferences file, then it doesn't  "travel" with the INO file. Better to use, as originally suggested, "meta-comments" in the INO file. Yes, one can have multiple files in a folder, includes, data, etc... but many users find a single, simple, INO file posted on the web and copy it in. Yes, Arduino code is very portable, but it's best for the IDE to at least _try_ to select the right one for the sketch when the author bothered to specify it. Keep in mind, the vast majority of INO files will not specify the chip. This is primary for those poor souls who are maintaining a large set of programs over a range of devices and are tired of switching. And also to help out new users who don't understand that a sketch might be for a specific device when the author is kind enough to hint that in the INO file. 

2. In any case, the IDE should never assume an action. Upon opening the file, and parsing meta-comments, it should _offer_ to make those changes for the user. The user can accept or reject the offer. If the user rejects the offer, no changes are made. The user may just be browsing the code. 

3. Keeping a simple format in the INO file means it also works as documentation for the human. E.g. 
`//  Tools / Board / Arduino Nano`
can be read by the IDE and also by a human. 

4. As one menu change may affect the availability of others, they should be processed in order. e.g.
`// Tools / Board / Node MCU 1.0 (ESP-12E Module)`
`// Tools / Flash Size / 4M (1M SPIFFS)`

5. As humans hate to type, it might be good to have a menu option that records the current menu settings into the top of the .INO or updates the existing settings. 

Things I'm less sure about:
- When to trigger the parsing? File / open only? File / New, Edit / Paste? Upon any change to the file, but only if the current menu settings don't match the parsed meta-comments? I would lean towards the latter, but that does mean that will be parsed on every build. That isn't too much of a burden is it?

- Which items should be included / excluded? Obviously, the File, Edit, Sketch, and Help menus should be excluded. Tools / Port and / Upload speed should _probably_ be excluded. Perhaps just start with everything on the Tools menus and tune in later releases?

- What to do if a board isn't available. Actually, that's pretty obvious, just error out. But the point is you have to look for that and interact with the user about it.

**Comment by timkoers** [comment]
Sure! I'll work in my own branch for now and place a pull request once the feature is finished

**Comment by Avamander** [comment]
@timkoers 
Just mentioning, compatibility with doxygen would be really nice (so that the generated documentation actually also mentions the sketchs' intended boards and etc.), it would be a good idea to even contact the doxygen developer if you aren't that familiar with doxygen yourself.

**Comment by timkoers** [comment]
Sure!
What I could do is use the doxygen syntax formatting so that doxygen is able to generate the data out of the box

**Comment by timkoers** [comment]
Can someone reopen this issue and add the label `in progress` to this issue?

**Comment by timkoers** [comment]
Just a little update, I managed to get it working!
The syntax is as following:

`
/** Arduino IDE Board Tool details
 * Board: Node32s
 * Flash Frequency: 80MHz
 * Upload Speed: 921600
*/`

**Comment by JamesNewton** [comment]
Love it!

Is it looking for the exact text "Arduino IDE Board Tool details"? If so, may I suggest something simpler and less likely to be incorrectly entered? E.g. "Settings" or "Tools Menu" or even "Arduino Tools Menu Settings". I'm trying to come up with things that are unlikely to be misspelled or typoed. E.g. You can look up and see the word "Arduino" in the ide so that's a good choice. And it is the "Tools Menu" so that's easy to remember. Probably "Arduino Tools Menu" is best. It tells you exactly where it's going to have effect.

**Comment by timkoers** [comment]
Sure! It is a single line of code to edit :-)
I'll add the header automatically when a new file is created and warn the user if there is a settings mismatch when loading the file.

Should I create a dialog that notifies the user that the settings have been loaded from the header?

Sent from Nine<http://www.9folders.com/>
________________________________
Van: JamesNewton <notifications@github.com>
verzonden: maandag 19 maart 2018 10:51 p.m.
Aan: arduino/Arduino
Cc: timkoers; Mention
Onderwerp: Re: [arduino/Arduino] Autoselection of device (#1481)


Love it!

Is it looking for the exact text "Arduino IDE Board Tool details"? If so, may I suggest something simpler and less likely to be incorrectly entered? E.g. "Settings" or "Tools Menu" or even "Arduino Tools Menu Settings". I'm trying to come up with things that are unlikely to be misspelled or typoed. E.g. You can look up and see the word "Arduino" in the ide so that's a good choice. And it is the "Tools Menu" so that's easy to remember. Probably "Arduino Tools Menu" is best. It tells you exactly where it's going to have effect.

**Comment by facchinm** [comment]
@timkoers thanks for your effort! I took a look at https://github.com/timkoers/Arduino/tree/board-autoswitch and the method is totally convincing. My only concern is that it's totally tied to Java IDE, so if you compile the same sketch using plain `arduino-builder` it doesn't apply the in-sketch configuration. 
It could also be a feature, so the plain builder only considers the fqbn provided explicitly.

**Comment by timkoers** [comment]
No problems! I’ll try to convert the methods into the arduino-builder (that is the arduino-core right?).
I’ll try to fix more issues like these, so it would be usefull if I can manage the issues of the Arduino repo too.

Tim

Van: Martino Facchin [mailto:notifications@github.com]
Verzonden: dinsdag 20 maart 2018 12:01
Aan: arduino/Arduino <Arduino@noreply.github.com>
CC: timkoers <tim.koers@live.nl>; Mention <mention@noreply.github.com>
Onderwerp: Re: [arduino/Arduino] Autoselection of device (#1481)


@timkoers<https://github.com/timkoers> thanks for your effort! I took a look at https://github.com/timkoers/Arduino/tree/board-autoswitch and the method is totally convincing. My only concern is that it's totally tied to Java IDE, so if you compile the same sketch using plain arduino-builder it doesn't apply the in-sketch configuration.
It could also be a feature, so the plain builder only considers the fqbn provided explicitly.

**Comment by per1234** [comment]
arduino-builder is here:
https://github.com/arduino/arduino-builder

**Comment by per1234** [comment]
It would be nice if the Arduino Web Editor and the Arduino IDE could use the same system for this. I'd like to see the functionality of the two diverge the least amount possible. That could mean either implementing the system added to the IDE code in the web editor or vice versa. It's not clear what system the Arduino Web Editor uses to store the sketch metadata. Making the changes in arduino-builder rather than the IDE would facilitate that goal since that tool is also used by Arduino Web Editor.

>It could also be a feature, so the plain builder only considers the fqbn provided explicitly.

I think if arduino-builder was made to use the sketch properties when no fqbn was specified but override those properties when one is specified then that would provide this feature and more.

Doing it in arduino-builder could lead to code duplication since the IDE must make the correct menu selections according to the sketch properties. The solution I see is for the IDE to run `arduino-builder -dump-prefs` on the sketch then make the correct menu selections according to arduino-builder's output.

**Comment by timkoers** [comment]
You mean code duplication as in, the Arduino IDE reads and parses the settings and the arduino-builder reads and parses the settings? 

I tried to execute the arduino-builder command, but it is not letting me. Is the output of the -dump-prefs command documented somewhere?

So to get confusion out of the way, I need to change the arduino-builder to do the things that I've programmed the IDE to do (read the header and select the right boards), and I need to reprogram the IDE to read the settings from the arduino-builder -dump-prefs command.

How would we handle the opening/reloading/closing of the file, should I execute the arduino-builder everytime?

Many thanks

**Comment by per1234** [comment]
>You mean code duplication as in, the Arduino IDE reads and parses the settings and the arduino-builder reads and parses the settings?

Exactly. It seems silly to have that same code in two places and need to maintain it.

>Is the output of the -dump-prefs command documented somewhere?

Well, somewhat:
https://github.com/arduino/arduino-builder/blob/master/README.md#usage
Or just run arduino-builder with no options for the same output.

>So to get confusion out of the way, I need to change the arduino-builder to do the things that I've programmed the IDE to do (read the header and select the right boards), and I need to reprogram the IDE to read the settings from the arduino-builder -dump-prefs command.

>How would we handle the opening/reloading/closing of the file, should I execute the arduino-builder everytime?

That was my suggestion. You should be aware that I don't really know what I'm talking about and have no official status in making decisions. So don't misinterpret what I say as coming from a member of the Arduino development team (such as facchinm).

**Comment by facchinm** [comment]
The direction is to remove as much logic as possible from the Java IDE, making it more or less an interface to cli tools. This will make it easier to share the tools between all "unofficial" IDEs (like VSCode/Atom) so the resulting behaviour is the same across all platforms. In this vision, "Add build settings to .INO file" JMenuItem would call an "endpoint"  provided by the builder that takes care adding the settings to the actual file.
Of course there are problems there too, because the builder is only aware of files already written on disk.
I don't have a real solution about this, but we can discuss freely, starting from @timkoers code. The feature appears to be much needed these days so I'll support every effort for fixing it in a sane way.

**Comment by timkoers** [comment]
I took a look at the android-builder source code, I don't understand a single line of the Go language.
I have no problems learning another programing language, but I don't think this is the right way (as I'll probably make the code very inefficient)
@facchinm do you have more experience coding in Go?

**Comment by JamesNewton** [comment]
I think it's important to see that this feature is an IDE, and NOT a compiler feature. It is NOT for CLI users, but for those who have little experience, to help them use the IDE and understand which board / settings the INO file is for, and for people who use the IDE all the time and so don't enjoy the nice features of CLI make files, scripting, etc... 

> Should I create a dialog that notifies the user that the settings have been loaded from the header?

More than that, I would strongly suggest adding a dialog that _asks_ the user if they wish to have the settings loaded. Actually, it should first check to see if the settings need to be loaded (e.g. if the current settings are any different from those in the meta-comment) and then check to see if they *can* be loaded (e.g. if those options are available) and then ask the user if it *should* be loaded. Once the user agrees, having a confirmation dialog is probably unnecessary.

**Comment by timkoers** [comment]
I am doing the last part in my code (check if they can be loaded) first, because it makes no sense to ask the user if the IDE should load settings and then notify that the settings are not found xD
I haven't implemented the dialog yet, if you'd like to do it, please make an preference option that will allow the more advanced user to disable the IDE to ask if it should load the settings everytime

**Comment by adwsystems** [comment]
As I am working with several different Arduino boards (Pro Mini 3.3V and 5V, UNO, MEGA, Nano, and possibly Micro. This is a feature that would be GREATLY useful. It would make sure that when I open the sketch the board selected is aligned with the board being used. Currently I am working on a project that uses 1 Mega, 1 UNO, and several Nanos.

**Comment by timkoers** [comment]
I made the code to integrate this into the IDE, I've created a pull request that they haven't accepted yet. Right now a merge conflict has occurred, due to them not merging my code 🤣 and changing part of the original code, that I haven't edited in my files.

**Comment by AndreKR** [comment]
Seems that branch needs a rebase.

**Comment by lmihalkovic** [comment]
I think I will do the following

* create arduino.json
* have the following contents
```
{
  environments:[
    "__all__": {
       "config": {
           "shared1": "valShared1"
       }
    }
    "my wet dream 1": {
       "config": {
           "var1": "val1",  "var2": "val2",  "var3": "val3",  "var4": "val4"
       }
    }, 
    "but this can also do": {
       "config": {
           "bar1": "bal1",  "bar2": "bal2",  "bar3": "bal3"
       }
    }
  ]
}
```
* add a dropdown list (on the editor toolbar) with "my wet dream 1" and "but this can also do" and I can choose between the two
* add a editor for arduino.json (doing the same that atmel had done to Visual Studio)


THE DROPDOWN IS A MOCK (the rest is not)
<img width="555" alt="screen shot 2018-05-03 at 8 54 42 am" src="https://user-images.githubusercontent.com/1318077/39625227-d71b401a-4f9c-11e8-95db-60236650952f.png">

adding a -env "my wet dream 1" to the command line of builder would make it do the right thing ... the gods helping..

**Comment by lmihalkovic** [comment]
This is a more realistic example that I will be using for my project. I added UUIDs so that even if the conf gets renamed builder can use these names as a basis for the cache name it generates (syntax highlighting is all wrong because I havent tried to changed it yet - but the editor understands what that file is)

<img width="708" alt="screen shot 2018-05-04 at 3 03 43 pm" src="https://user-images.githubusercontent.com/1318077/39629431-736d5e94-4fac-11e8-9a4a-c71c6242e07b.png">

**Comment by mirage335** [comment]
Solved.

ArduinoUbiquitous looks in the sketch directory for an "ops" file, importing it into its environment. Any shell commands, including preferences configuration, such as board type, can be hooked and forced from there.
https://github.com/mirage335/arduinoUbiquitous
https://github.com/mirage335/arduinoUbiquitous/blob/master/_local/ops
https://github.com/mirage335/arduinoUbiquitous/blob/master/_prog/core.sh

The framework also contains applications, like Arduino, to use arbitrary home folders, and provides a structure to maintain 'splice' instructions against existing software. Such functionality can be used to maintain a a statically compiled recent OpenOCD build, a custom Arduino hardware core, or a custom blockly visual programming environment.
https://github.com/mirage335/openocd-static
https://github.com/mirage335/SigBlockly-mod
https://github.com/mirage335/SigBlockly

Extremely flexible virtualization features allow the entire development toolchain to be launched, from the host filesystem, in a virtual machine image which can simultaneously be run under ChRoot, Docker, QEMU, and VirtualBox. Multiple toolchains can be contained to a "lab" set of directories, all Ubiquitous Bash forks using git submodules.
https://rawgit.com/mirage335/ubiquitous_bash/master/USAGE.html
https://github.com/mirage335/ubiquitous_bash/blob/master/lab

Atom is integrated. A general purpose Arduino debugging IDE is on the way.

**Comment by mirage335** [comment]
It is a huge wrapper bash shell script for ArduinoIDE, which includes functionality to open a text file named "ops" in the same directory as the sketch, and run commands from that file before launching the IDE. This wrapper script is also extensively broken down into functions, some of which are meant to be overridden by re-defining with this "ops" file, effectively 'hooking' some code to be run at various times in the launch sequence.

So you could put this, in a file named "ops", in the same directory as "Blink.ino".
```
#Uncomment to force. Consider configuring through GUI instead.
_arduino_prepare_board() {
	#true
	_arduino_set_board_zero_native
}
```
Which runs the "_arduino_set_board_zero_native" function, included with the wrapper, to run before the IDE, forcing the board type to be set to Arduino Zero (Native).
Then you would run something like this.
"../arduinoUbiquitous/_arduino ./Blink.ino"

Arduino itself is just installed to "_local/h/arduino-1.8.5" or similar. This is already included in the repository. The "_local/h" folder is used as the user's home folder when launching the IDE, though symlinks are used to allow the ArduinoIDE to use its own internal "portable" mode with the same files.

**Comment by JamesNewton** [comment]
@mirage335 a couple of questions:

1. "bash shell script" so that means it only works on Linux systems?

2. Can you put a script somewhere other than the project directory, which will run when a new .ino files is opened, which will look through the commends in the .ino file to see if there are comments which direct the IDE to select a specific board / board settings / etc..., popup a dialog to confirm that with the user and then finally, actually make those setting changes?

I ask #2 because having a second file in each project folder is exactly what I don't want. Instead:
https://github.com/arduino/Arduino/issues/1481#issuecomment-374064524

**Comment by mirage335** [comment]
#1 No. Technically Linux is _most_ supported, because it is what I am using, but the underlying scripts have been used under Cygwin, autodetection of MSW executables is possible with the architecture, and OSX support in theory partially works already. That said, Linux is both open-source and by far the most convenient platform for development, so I expect the most serious developers to use it when possible.

#2 Yes. This could actually be done on a per-project basis. You would only need to place an "ops" file in the wrapper script's own directory, overloading the "_arduino_prepare_board" function with an implementation to read the sketch file for this information, and install it. Best way to do that might be to allow raw shell commands to be included in the file. Such an implementation could also replace the current placeholder under "core.sh".

**Comment by per1234** [comment]
This issue tracker is for bug reports or feature requests for official Arduino software. I'm sure arduinoUbiquitous is really great but this is not an appropriate place to promote or discuss a 3rd party project.

**Comment by mirage335** [comment]
Correct. The discussion as to how to use arduinoUbiquitous, or any issues it has, should be taken to its issue tracker.

However, my intent in mentioning it here was to point out a solution - specific to Arduino itself - to an essential feature request left unaddressed since 2013. Having done so with an external framework may also imply that the original feature request is not within scope of the Arduino project itself.

**Comment by lmihalkovic** [comment]
IMHO this is just one of the many hackish way of addressing the question of board selection. I have no doubt that it would work, but there are times when no solution is better than a poorly devised one, because it leaves the hope alive that there might one day be a good one.

**Comment by mirage335** [comment]
UNIX philosophy. Do One Thing and Do It Well.

> Arduino can be used to develop stand-alone interactive objects or can be connected to software on your computer (e.g. Flash, Processing and MaxMSP).

ArduinoIDE compiles/uploads code, and includes some libraries. Beyond that, it is the growing ecosystem that makes it powerful. Shields, machine control panels, g-code generators, web interfaces, etc.

For the Arduino project itself, I wonder if embedding a specialized, probably java based, "meta-comment" processor is considered gold plating. Some challenging extensibility features could be required to address user's needs. Also, a debugger would seem to be in a similar category of desired functionality, as an example of how difficult this entire category of functionality could be to integrate directly into Arduino.

Meanwhile, there seems to be plenty of ongoing bug fixing and release cycle work going on, directly related to what Arduino does really well. Focusing on stabilizing Arduino itself, maybe making it even more modular (Arduino Builder looks cool by the way), and maybe supporting interfaces third party software can use, might do much more for power users in the long term.

**Comment by Avamander** [comment]
@mirage335 
It's an IDE, the whole point is to gather up a lot of tools and make them usable.

**Comment by timkoers** [comment]
How would a strong and robust board header/board definition look like when it's up to you?
I think we should discuss about how we are going to implement a solution instead of what we shouldn't do?

By the way, I already coded a solution that looks like this in your Arduino file

[Issue 1481](https://github.com/arduino/Arduino/issues/1481#issuecomment-374366404)

**Comment by mirage335** [comment]
At least arbitrary setting of any Arduino preference, including programmer. Overriding the libraries directory would be useful too. Maybe some sort of arbitrary code execution, search and replace, as well.

@Avamander An IDE of highly limited scope. Just on account of not having a debugger, forcing specific libraries for various projects, and keeping a complete compiler with git submodules for reference, Arduino by itself still will not be usable for me.

**Comment by timkoers** [comment]
@mirrage335<https://github.com/mirrage335> can you be more specific when it comes to “At least arbitrary setting of any Arduino preference, including programmer. Overriding the libraries directory would be useful too. Maybe some sort of arbitrary code execution, search and replace, as well.“

Van: mirage335 [mailto:notifications@github.com]
Verzonden: vrijdag 1 juni 2018 23:00
Aan: arduino/Arduino <Arduino@noreply.github.com>
CC: timkoers <tim.koers@live.nl>; Mention <mention@noreply.github.com>
Onderwerp: Re: [arduino/Arduino] Autoselection of device (#1481)


At least arbitrary setting of any Arduino preference, including programmer. Overriding the libraries directory would be useful too. Maybe some sort of arbitrary code execution, search and replace, as well.

@Avamander<https://github.com/Avamander> An IDE of highly limited scope. Just on account of not having a debugger, forcing specific libraries for various projects, and keeping a complete compiler with git submodules for reference, Arduino by itself still will not be usable for me.

**Comment by mirage335** [comment]
Everything in preferences.txt . Equivalent to running this from the meta-comment.
`--save-prefs --pref programmer=arduino:sam_ice`

**Comment by Avamander** [comment]
@timkoers Your suggestion is exactly what I'd like to see.

@mirage335 You have the C(++) preprocessor for most of what you've described. Board selection as such can't be done by it and that's why it's discussed here. Editing an additional file loses in both clarity, UX and usability and thus in my opinion can't be the solution.

> Maybe some sort of arbitrary code execution, search and replace, as well.

That's a really horrible idea from security perspective and must not be implemented.

**Comment by mirage335** [comment]
C Preprocessor cannot set the programmer type. When dealing with custom cores, having a programmer set to something from another core seems to cause build failures in the ArduinoIDE. Moreover, setting board type cannot force a huge 3d printer firmware codebase to be compiled against libraries in the local directory, which is my biggest nuisance with Arduino. Asking me to install anything to a global directory because some specific microcontroller firmware version needs it is completely unacceptable. I would rather set up a dedicated virtual machine, but should not have to.

`That's a really horrible idea from security perspective and must not be implemented.`
To be clear, I am suggesting the ability to manipulate variables in the ArduinoIDE, not to run code on the host computer. But while even that might be unwise to enable by default, "must not exist' is a strong position to take. Arduino firmware developers are working on code that will at least have root access to the microcontroller itself. In the context of Internet of Things, machine control, and industrial automation, the security issues are already there.

**Comment by timkoers** [comment]
@Avamander it's [here](https://github.com/arduino/Arduino/pull/7349)!

**Comment by mike-s123** [comment]
This has been open for too long. It shouldn't be that hard.

1) Allow different boards when there are multiple sketches open. It's really difficult to try to work with different boards at the same time.

2) Allow save/load of board setups. It shouldn't directly be part of an .ino, but should be saved in a sketchbook (sketch folder), so it's loaded automatically when opening a project and also gets archived with the sketch.

**Comment by atrent** [comment]
I'd like to "+1" to this RFE.
A way to save board (and related settings) parsable info into the code or as a side file would be very useful.

**Comment by mirage335** [comment]
+1
Again.

**Comment by GarryG** [comment]
How is it, that after several years of people noticing that there has to be some sort of persistent information in a project as to what the compile settings are, nothing is happening with this?
Most compiler environments allow setting the options and keeping them WITH THE PROJECT, usually in a make-file or similar ... not having any way of setting the global compiler settings apart from the IDE itself, and it ignoring the fact that it is being used for dozens of boards, appears pretty ignorant ...
I've not had much of a problem before, as I usually only compiled Wemos D1 compatible boards (apart from the odd ESP01 here and there), but have now gotten some Arduino Nano boards, which led to me noticing that the settings have to be changed every f***ing time I switch from one project to another ... really, this can't be that hard ... just scan the first and last 10-50 lines for comments, have some keywords and translate the stuff you found to the IDE settings ... this would also allow for the comment to be automagically added/updated to the source INO file on changing the settings through the IDE ...
IMHO, this is a 100-200 line code update for the Arduino IDE ... (plus, I assume, additional lines to allow for multiple projects open with different settings ... this also is a must-have ...)

**Comment by MarkusNaeher** [comment]
For the time being, I can suggest a workaround that makes switching the board at least easier:

Edit boards.txt (the exact location depends on your OS) and comment out all boards you don't use.
If you've added additional board types like ESP8266, they come with their own boards.txt. I suggest to edit them also. After that, I only have a small number of options in the board selection that's easy to handle.

**Comment by tschundler** [comment]
Another (interim?) solution is to use VSCode as a frontend to Arduino IDE - this plugin for VSCode from MSFT: https://marketplace.visualstudio.com/items?itemName=vsciot-vscode.vscode-arduino saves board settings with a project. (You still use Arduino IDE to manage boards and libraries with that)

**Comment by JamesNewton** [comment]
I have no idea why I'm bothering to burn cycles on this again. I guess like @GarryG it just completely fry's me that after all these years, still there is no way to save board settings in the code. 

Let me tell you what the #1 issue is that I run into teaching kids and adults to program Arduinos: They get the IDE installed (no small feat), They get the USB driver / connection for whatever Arduino compatible device they have working (no small feat), then they select some example code, e.g. "blink" and try to program it. It fails. Because they have the wrong device selected. And the error message does NOT say anything like "go to Tools / Board..." and select the board you have. And the example programs don't mention needing to do that either. 

I could be wrong, but I seriously think that if you surveyed new users of the Arduino IDE, the point at which they most often give up and walk away is not knowing how to select the correct device for the sketch. 

And then there are sketches which NEED to be on a specific board. E.g. anything for the ESP-8266 / ESP-32 that needs WiFi. Or code that needs the speed of the Zero. There should be a /standard/ way for the people who write and share sketches to include that requirement IN the .INO file. Not in a separate file, not in a "project", because people starting out with these things don't even understand projects. They just copy and paste in an .INO file. When they compile it for the first time, the IDE should popup a window saying "Hey, what board is this for? I notice in the .INO a comment that looks like I should set the Tools / Board... to xxxx... You want I should do that for ya boss?"

And then there are professional developers who are just tired of having to remember all the bloody menu settings for each project as they switch from one to the other. I'm currently working on two different customer projects, both with the requirement that the work be done in the Arduino IDE so that they can take over maintenance when I'm done. One is on the ESP-32, the other is Arduino Nano but with the 168. So there are multiple settings per projects in each case. Very, very, very annoying. I want to be able to put, in the main INO file, a set of comments which represent the menu settings. And when I open and compile that project for the first time in one session, it should ask me if I want to make those menu changes.

Very simple, very useful, very clean.

**Comment by timkoers** [comment]
> Very simple, very useful, very clean.
>
Amen, if you really need a solution right now, you can clone my [branch ](https://github.com/timkoers/Arduino/tree/board-autoswitch)

It has got a little comment I talked about in this issue that specifies which board and settings are used. 

I am willing to redo this feature, but I need guarantee that the PR will be merged :-)

**Comment by Avamander** [comment]
I have waited for this feature for nearly five years now :rofl:, would be nice to have someone from the official dev team would back this and help get this feature merged.

**Comment by atrent** [comment]
[**PARTIAL** solution] is to save a copy of your **working** 'preferences.txt' (it's in ~/.arduino15) inside your (hopefully git) repository so that you have a future reference handy.

**Comment by tschundler** [comment]
@timkoers, have you submitted a pull request for that? Was it rejected? (If so, why?)

**Comment by timkoers** [comment]
@tschundler I submitted a pull request (https://github.com/arduino/Arduino/pull/7349) that has the feature that we are talking about implemented. I don't think it was rejected, it was not accepted.

**Comment by JamesNewton** [comment]
Yeah, at first I thought the developers felt this was a bad idea, based on the really stupid argumentative reactions to my issue: #5308 . I ended up closing it because I just couldn't deal. But after some time, I've realized that those people were not the actual developers; just bots who can't reason and like to argue. The only active commits to Arduino come from Christian and Martino and they have completely ignored this. Either they are apathetic or paralyzed. It may be they are afraid to change the UI at all. @facchinm (Martino) seemed to appreciate the idea "...the method is totally convincing" but then had some incomprehensible (to me) objection regarding the IDE vs the CLI? I tried to address that, but was, again, ignored. This is NOT (wish I had a dime for the number of times I've said this) a compiler or CLI issue. It's an /interaction/ between the user of the IDE and the IDE. Like syntax highlighting or autocomplete. It MUST be in the IDE code, and is not needed in the CLI code. Perhaps his issue was wanting some sort of defines that would drop the code out of the CLI compile? Or I'm confused what the holdup actually is.

**Comment by per1234** [comment]
@JamesNewton the tone you're using is unproductive and will result in this thread being locked if it continues. If you and others are truly interested in participating in improving the Arduino software then take a different approach to your communication.

I'm surprised you can't see the need for this feature to be integrated at the level of the CLI. The association should work the same whether the user is using the IDE or the CLI. The Arduino IDE code is moving in the direction of being only for the GUI. All the underlying functionality is provided by the [arduino-cli](https://github.com/arduino/arduino-cli) tool, which is shared between the Arduino IDE, Arduino Web Editor, Arduino Pro IDE, and 3rd party software. So by adding a feature to arduino-cli, it becomes available to all the software that uses that tool.

arduino-cli already supports the use of a metadata file (sketch.json) stored in the sketch folder to associate a board and port (though I'm not sure the port aspect is fully working yet). There is a dedicated CLI command for this: `arduino-cli board attach`. This feature is already available in the Arduino Web Editor. I can't say whether there are any plans to do the same for the Arduino IDE, but it seems clear to me that if it were to be done this would be the mechanism used.

**Comment by JamesNewton** [comment]
@per1234 Oh, I'm terribly sorry if my tone came over wrong. I seriously meant no offense to you or anyone else. I guess my frustration with prior interactions with non-contributors is boiling over here and I apologize for that. Since you actually have contributed to the project, I'll take the time to try to explain further:

Perhaps the word "interaction" is confusing. Let me clarify: This must go two ways. If you put something in the comments of the source .ino file which /always/ causes a change to settings, that violates least user surprise and breaks the CLI. If you pop up a dialog and /ask/ what the user wants, surprise is mitigated. 

Popping up a dialog is generally not done in CLI's, at least as far as I've seen. Would you want the CLI to stop mid way and ask the user about which board to target? For the CLI, a separate (optional) settings file, or the port / board association makes perfect sense. Interaction doesn't happen in CLI's except via the command line. The GUI IDE can write a settings file or, more obviously, change the command line options when invoking the compiler. This is a GUI IDE feature only.

I guess it could be valuable for the CLI to issue a warning if the comment indicates one board and you're command line settings are compiling for another? But more than that would be problematic.

Now, you could argue that the port / board association or a separate settings file makes sense in the GUI, but it has a serious limitation: The user must know about it. As I've indicated above, most beginning users are not aware of the need to select a board. Even if they understand the need to select for the board they own, the .INO that they pasted in often needs a specific board and this is not noted in the sketch which is all they have. An easy way for authors to add that to the file, in the comments so it doesn't break the CLI, and have the GUI IDE point it out to you, is a help.

Interactions are typically associated with the GUI. We are specifically asking for a change to the GUI IDE ONLY for /interaction/ with the user. Again: It's like syntax highlighting or autocomplete. It simply works WITH the user /interactively/ to help a new user select the correct board for the project. 

Interactive. GUI only. 

Like autocomplete, or syntax highlighting, or copy and paste, or typing in the .INO file. 

Very simple: The GUI IDE looks at the comments, finds a matching block, sees that the menus don't match and this is the first time this session, so it asks the user if they want the menus updated from the comment. The user can say "Yes" or "No" and that's the end of the interaction. If they said yes, the menus are updated to match the INO.

**Comment by timkoers** [comment]
@JamesNewton +100000! 
That is exactly what my PR (https://github.com/arduino/Arduino/pull/7349) does, except for the dialog to ask the user if they'd like to get the settings matched with the comments in the .INO file.

**Comment by tschundler** [comment]
Where is sketch.json documented? Quickly search Google found example of it, but no docs. It sounds like maybe a solution for most(?) cases.

**Comment by per1234** [comment]
>If you put something in the comments of the source .ino file which /always/ causes a change to settings, that violates least user surprise and breaks the CLI.

But it doesn't always cause a change. Let's say you do this:
```
arduino-cli board attach arduino:avr:nano Foo
```
Now the default board for that sketch is the Nano. So if you do this:
```
arduino-cli compile Foo
```
it will compile for the Nano. However, if you do this:
```
arduino-cli compile --fqbn arduino:avr:uno Foo
```
then it will compile for the Uno. The only time you wouldn't use the `--fqbn` option is if you specifically wanted to use the default board associated with the sketch. So arduino-cli compiling for Nano in that case is not surprising to the user. Therefore there is no need for interaction in the CLI because the user is specifying exactly how they want arduino-cli to behave in advance in the command.

>I guess it could be valuable for the CLI to issue a warning if the comment indicates one board and you're command line settings are compiling for another?

No need, since the command line settings, if present, override the default settings in sketch.json.

>Now, you could argue that the port / board association or a separate settings file makes sense in the GUI, but it has a serious limitation: The user must know about it.

Not if the IDE has a GUI interface that allows them to set the associations. They don't need to know the mechanism behind how that information is stored. This similar to the Arduino IDE's **File > Preferences** interface. The preferences are stored to the preferences.txt file, but the user doesn't need to know that. They just set their preferences via the GUI. This is similar to how arduino-cli has a `arduino-cli board attach` command. Even the CLI user doesn't need to know about sketch.json.

>Where is sketch.json documented?

I don't think it has been documented. It would be good to document this somewhere in the arduino-cli GitHub repository for people who might want to use that file directly. However, arduino-cli's command line help does document the `board attach` command, which is the normal user's interface to that file:
```
>arduino-cli help board attach
Attaches a sketch to a board.

Usage:
  arduino-cli board attach <port>|<FQBN> [sketchPath] [flags]

Examples:
  C:\Users\per\go\bin\arduino-cli.exe board attach serial:///dev/tty/ACM0
  C:\Users\per\go\bin\arduino-cli.exe board attach serial:///dev/tty/ACM0 HelloWorld
  C:\Users\per\go\bin\arduino-cli.exe board attach arduino:samd:mkr1000

Flags:
  -h, --help             help for attach
      --timeout string   The timeout of the search of connected devices, try to high it if your board is not found (e.g. to 10s). (default "5s")

Global Flags:
      --additional-urls strings   Additional URLs for the board manager.
      --config-file string        The custom config file (if not specified the default will be used).
      --format string             The output format, can be [text|json]. (default "text")
      --log-file string           Path to the file where logs will be written.
      --log-format string         The output format for the logs, can be [text|json]. (default "text")
      --log-level string          Messages with this level and above will be logged. (default "info")
  -v, --verbose                   Print the logs on the standard output.
```

**Comment by dsyleixa** [comment]
IMO the most simple way would be to hardcode-config the targeted board type by configuration in Sketch (e.g. #pragma config...) such as
#pragma config board="Nano"
#pragma config board="NODE_MCU 1.0"
#pragma config board="Adafruit Feather ESP32"
